### PR TITLE
Move resource updates to different thread.

### PR DESCRIPTION
**Pull request by fibann**
Also fix a few bugs with resource timers.

### CODE REVIEW

File: libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs [hunk 1]
```diff
@@ -724,25 +781,31 @@ private void TaskDispose(DiscoveryTask task, CategoryInfo info)
...
+            // Use int since UWP does not implement long ctor.
+            int deltaMsInt;
+            if (expiryTime == DateTime.MaxValue)
             {
-                var deltaMs = (long)DateTime.UtcNow.Subtract(expiryTime).TotalMilliseconds;
+                deltaMsInt = Timeout.Infinite;
+            }
+            else
+            {
+                var deltaMs = (long)expiryTime.Subtract(DateTime.UtcNow).TotalMilliseconds;
```
Comment (fibann) [inline libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs]:
```
- var deltaMs = (long)DateTime.UtcNow.Subtract(expiryTime).TotalMilliseconds;
+ var deltaMs = (long)expiryTime.Subtract(DateTime.UtcNow).TotalMilliseconds;
```
This was a bit embarassing...

File: libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs [hunk 2]
```diff
@@ -571,13 +571,57 @@ class Client
...
+                var handles = new WaitHandle[] { timerExpired, updateAvailable_ };
+                var tasksUpdated = new List<DiscoveryTask>();
+                var token = updateCts_.Token;
+                using (var updateTimer = new Timer(_ => { timerExpired.Set(); }))
+                {
+                    // Stop the wait if cancellation is requested.
+                    token.Register(() => updateAvailable_.Set());
+                    while (true)
+                    {
+                        WaitHandle.WaitAll(handles);
```
Comment (stephenatwork) [inline libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs]: Don't you need to pass the token here?

Comment (fibann) [reply libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs]: No, this is Wait**All**. If the cancellation is issued I wake up this call by calling `updateAvailable_.Set()` (line 596), but I don't want to wait for cancellation every time 😄 

Will add comments to clarify this.

Comment (stephenatwork) [reply libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs]: Talked IRL - WaitAll(Task[], CancellationToken) looks like what we need

Comment (fibann) [reply libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs]: Re-talked IRL - WaitHandle.WaitAll doesn't take a CancellationToken (for some reason).

We established that the minimum interval is not that useful and removed it.

**Pull request merged**